rfd/0080-yubikey-piv.md

```diff
@@ -0,0 +1,272 @@
+
+## Details
+
+Yubikeys (Series 5 and on) have a built in smart card which provides PIV compatibility, which we can interact with using the [go-piv library](https://github.com/go-piv/piv-go).
+
+### yubikey login
+
+We'll add the `tsh` flag `--yubikey/-y` so that users can login with the new yubikey login flow. The majority of the login flow is the same, except the key generation and certificate signing process.
+
+#### Generate yubikey key pair and certificates
+
+When `tsh login` would usually generate a new private key in memory, it will instead use [`piv.GenerateKey`](https://pkg.go.dev/github.com/go-piv/piv-go/piv#YubiKey.GenerateKey) to generate a new key pair directly on the yubikey's PIV Authentication slot (`9a`). The certificate signing process will continue as usual, using the yubikey's public key, and the certs will be stored in `~/.tsh/keys/proxy` as usual.
+
+Neither the yubikey's public or private key will be stored, as the private key can't be exported and we can later access the public key by checking the yubikey's [slot cert](https://pkg.go.dev/github.com/go-piv/piv-go/piv#YubiKey.Attest).
+
+##### Yubikey PIN/Touch policies
+
+When we generate a new private key, we can provide it with a custom Touch and PIN policy. These policies will be enforced whenever the private key is used for signing or decrypting, which may occur multiple times for each `tsh` command.
+
+The available PIN policies are:
+ - `never`: never prompt the user for PIN
+ - `always`: always prompt the user for PIN
+ - `once`: prompt the user for PIN just once per session, where a session is an open connection to the yubikey (needs verification)
+   - Due to the current keystore implementation reloading keys multiple times for every command, this may still lead to the user being prompted for PIN multiple times in each tsh call. This can potentially be reworked, but it will most likely not be fixed in the initial implementation.
+
+The available Touch policies are:
+ - `never`: never prompt the user for touch
+ - `always`: always prompt the user for touch
+ - `cached`: cache user's touch for 15 seconds before prompting them for it again
+   - This should result in each `tsh` command requiring just one touch
+   - We can try to control when the user is prompted for touch by doing a no-op signature at the start of a `tsh` command. However, if the cached touch runs out during the command, the user may be prompted later in the command execution than desired.
+
+The user can provide `--yubikey-pin-policy=<never|always|once>` and `--yubikey-touch-policy=<never|always|cached>` to specify their desired policies. They will both default to `never`.
```

Comment:
> They will both default to `never`

Doesn't that defeat the purpose of using Yubikey storage? What's to stop another process from using an unguarded key?

I would guess that pin:once and touch:cached seem like good defaults. (Caveat: I never played with yubi PIV.)

Reply:
The idea is that `never` will give the best UX, simply requiring the user to have the yubikey connected. The private key is still securely in the yubikey and cannot be exported, so its still a step safer than plain `tsh login`.

But I agree that `once` and `cached` would also be reasonable defaults. Leaving this unresolved to get a third opinion.

Reply:
I expect never/never to allow any process to use the keys without user interaction, is that correct? If yes it's better than on disk, but not by much, especially for yubikeys that we never unplug.

Reply:
Adding to this, considering what the default PIN is, a touch policy would go a long way in safeguarding the keys.

Reply:
Yeah I think you're right, updated (in the `tsh` flags section)